datastore/adapter/keytransform.py

```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
 		datastore.abc.ObjectAdapter[T_co, T_co]
 ):
 	__slots__ = ("key_transform_fn", "nest_depth", "nest_length", "nest_keyfn")
+
+
+class _FlatFSAdapter(_Adapter[DS, RT, RV], typing.Generic[DS, RT, RV]):
+	"""Represents a simple DatastoreAdapter that shards/namespaces incoming keys.
+
+	Incoming keys are sharded into namespaces according to specifications
+	in a `SHARDING` file, which `prefix`, `suffix`, and `next-to-last`
+	sharding functions available. For example:
+	
+	Implementation Notes:
+		* The `default_prefix` is "/repo/flatfs/shard/"
+		* The only version supported is "v1"
+		* function_name must be `prefix`, `suffix`, or `next-to-last`
+
+	"""
+
+	prefix: str = "/repo/flatfs/shard/"
+	_default_sharding_key: datastore.Key = datastore.Key("SHARDING")
+	_default_sharding_func: str = "v1/prefix/2"
+
+	@classmethod
+	async def create(
+			cls: typing.Type[T],
+			child_datastore: DS,
+			*args,
+			sharding_key: datastore.Key = None,
+			default_sharding_func: str = None,
+			**kwargs
+	) -> T:
+		"""Initializes _FlatFSAdapter with sharding function.
+
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to initialize an `Adapter`
+		*args
+			Variable length list argument to be passed to cls constructor
+		sharding_key
+			Key used to access the sharding function in the underlying Datastore
+		default_sharding_func
+			Default sharding func to use if one is not found in the datastore
+		"""
+		# Get the variables needed to parse the sharding function
+		sharding_key = (
+			sharding_key
+			if sharding_key is not None
+			else cls._default_sharding_key  # type: ignore[attr-defined] # noqa: F821
+		)
+		default_sharding_func = (
+			default_sharding_func
+			if default_sharding_func is not None
+			else f"{cls.prefix}{cls._default_sharding_func}"   # type: ignore[attr-defined] # noqa: F821
+		)
+		sharding_fn: KEY_TRANSFORM_T = await \
+			cls._parse_sharding_function(  # type: ignore[attr-defined] # noqa: F821
+				child_datastore, sharding_key, default_sharding_func
+			)
+
+		# Return an instantiated version of the class
+		return cls(
+			child_datastore, *args, key_transform=sharding_fn, **kwargs
+			)  # type: ignore[attr-defined] # noqa: F821
+
+	@staticmethod
+	def _prefix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).ljust(length, "_")[:length])
+	
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	@classmethod
+	async def _parse_sharding_function(
+			cls: typing.Type[T],
```

Comment:
Don't use `cls: typing.Type[T],` here, it will let you drop a bunch of mypy attr-defined comments in this method.